ml4gw/distributions.py

```diff
@@ -173,3 +178,177 @@ def rsample(self, sample_shape: torch.Size = None) -> Tensor:
+        self.grid_size = grid_size
+
+        z_grid = torch.linspace(0, z_grid_max, grid_size)
+
+        omega_l = 1 - omega_m
+        h = h0 * torch.sqrt(omega_m * (1 + z_grid) ** 3 + omega_l)
+
+        dz = z_grid[1] - z_grid[0]
+
+        # C is specfied in m/s, h0 in km/s/Mpc, so divide by 1000 to convert
+        comoving_dist_grid = torch.cumsum((C / h) * dz, dim=0) / 1000
+        # Compute luminosity distance from comoving distance
+        luminosity_dist_grid = comoving_dist_grid * (1 + z_grid)
+
+        if distance_type == "comoving_distance":
+            z_min, z_max = self._linear_interp_1d(
+                comoving_dist_grid, z_grid, Tensor([minimum, maximum])
+            )
+        elif distance_type == "redshift":
+            z_min, z_max = Tensor([minimum, maximum])
+        else:
+            z_min, z_max = self._linear_interp_1d(
+                luminosity_dist_grid, z_grid, Tensor([minimum, maximum])
+            )
+
+        if z_max > z_grid_max:
+            raise ValueError(
+                f"Maximum {distance_type} {maximum} "
+                f"exceeds given z_max {z_grid_max}."
+            )
+
+        # Restrict grids based on the minimum and maximum values
+        mask = (z_grid >= z_min) & (z_grid <= z_max)
+        z_grid = z_grid[mask]
+        comoving_dist_grid = comoving_dist_grid[mask]
+        luminosity_dist_grid = luminosity_dist_grid[mask]
+        h = h[mask]
+
+        # Compute the pdf and log_pdf from the comoving distance grid
+        dV_dz = comoving_dist_grid**2 / h
+        if source_frame_time:
+            dV_dz /= 1 + z_grid
+        pdf_grid = dV_dz / torch.sum(dV_dz * dz)
+
+        log_pdf = torch.log(pdf_grid)
+
+        # Compute change of variables factors for computing log_prob
+        dcomoving_dist_dz = torch.gradient(comoving_dist_grid, spacing=dz)
+        dluminosity_dist_dz = torch.gradient(luminosity_dist_grid, spacing=dz)
+
+        self.minimum = minimum
+        self.maximum = maximum
+        self.z_grid = z_grid
+        self.dz = dz
+        self.comoving_dist_grid = comoving_dist_grid
+        self.luminosity_dist_grid = luminosity_dist_grid
+        self.log_pdf = log_pdf
+        self.dz_dcomoving_dist = 1 / dcomoving_dist_dz[0]
+        self.dz_dluminosity_dist = 1 / dluminosity_dist_dz[0]
+
+    def _linear_interp_1d(self, x_grid, y_grid, x_query):
+        idx = torch.bucketize(x_query, x_grid, right=True)
+        idx = idx.clamp(min=1, max=len(x_grid) - 1)
+
+        x0 = x_grid[idx - 1]
+        x1 = x_grid[idx]
+        y0 = y_grid[idx - 1]
+        y1 = y_grid[idx]
+
+        # Factor of epsilon for stability
+        t = (x_query - x0) / (x1 - x0 + 1e-20)
+        return y0 + t * (y1 - y0)
+
+    def rsample(self, sample_shape: torch.Size = None) -> Tensor:
+        sample_shape = sample_shape or torch.Size()
+        u = torch.rand(sample_shape)
+
+        pdf = self.log_pdf.exp()
+        cdf = torch.cumsum(pdf * self.dz, dim=0)
```

Comment:
move this to the constructor. We can keep these as attributes and reuse them for every sample call.